File: src/apps/daemon/internal/networking/tlsdecoy/sni_spoof_engine_test.go
package tlsdecoy

import (
	"encoding/binary"
	"net"
	"testing"
)

func Test4TupleKey(t *testing.T) {
	srcIP := net.ParseIP("192.168.1.10")
	dstIP := net.ParseIP("104.16.1.1")
	key := NewConn4TupleKey(srcIP, 54321, dstIP, 443)

	if key.SrcPort != 54321 || key.DstPort != 443 {
		t.Fatalf("unexpected ports: %d -> %d", key.SrcPort, key.DstPort)
	}

	rev := key.Reverse()
	if rev.SrcPort != 443 || rev.DstPort != 54321 {
		t.Fatalf("unexpected reverse ports: %d -> %d", rev.SrcPort, rev.DstPort)
	}
	if rev.SrcIP != key.DstIP || rev.DstIP != key.SrcIP {
		t.Fatalf("reverse IPs do not match")
	}
}

func TestBuildFakeClientHello(t *testing.T) {
	var random, sessionID, keyShare [32]byte
	for i := 0; i < 32; i++ {
		random[i] = byte(i)
		sessionID[i] = byte(i + 32)
		keyShare[i] = byte(i + 64)
	}

	fakeSNI := "speedtest.net"
	ch, err := BuildFakeClientHello(fakeSNI, random, sessionID, keyShare)
	if err != nil {
		t.Fatalf("BuildFakeClientHello failed: %v", err)
	}

	// Canonical length must strictly equal 517 bytes
	if len(ch) != 517 {
		t.Fatalf("expected 517 bytes, got %d", len(ch))
	}

	// Test with 6-char SNI "mci.ir"
	ch2, err := BuildFakeClientHello("mci.ir", random, sessionID, keyShare)
	if err != nil {
		t.Fatalf("BuildFakeClientHello failed with mci.ir: %v", err)
	}
	if len(ch2) != 517 {
		t.Fatalf("expected 517 bytes for mci.ir, got %d", len(ch2))
	}
}

func TestBuildFakePayloadPacket(t *testing.T) {
	// Synthesize 40-byte IPv4 + TCP packet
	raw := make([]byte, 40)
	raw[0] = 0x45 // Version 4, IHL 5
	binary.BigEndian.PutUint16(raw[2:4], 40)
	binary.BigEndian.PutUint16(raw[4:6], 0x1234) // ID
	raw[8] = 64                                 // TTL
	raw[9] = 6                                  // TCP
	copy(raw[12:16], net.ParseIP("192.168.1.1").To4())
	copy(raw[16:20], net.ParseIP("104.16.1.1").To4())
	computeIPv4Checksum(raw[:20])

	binary.BigEndian.PutUint16(raw[20:22], 54321) // SrcPort
	binary.BigEndian.PutUint16(raw[22:24], 443)   // DstPort
	binary.BigEndian.PutUint32(raw[24:28], 1000)  // Seq
	binary.BigEndian.PutUint32(raw[28:32], 2000)  // Ack
	raw[32] = 0x50                                // DataOffset: 5 (20B)
	raw[33] = tcpFlagACK
	computeTCPChecksum(raw, 20)

	fakePayload := []byte("DECOY_PAYLOAD_TEST")
	fakeSeq := uint32(99000)
	fakePkt, err := BuildFakePayloadPacket(raw, fakePayload, fakeSeq)
	if err != nil {
		t.Fatalf("BuildFakePayloadPacket failed: %v", err)
	}

	if len(fakePkt) != 40+len(fakePayload) {
		t.Fatalf("expected length %d, got %d", 40+len(fakePayload), len(fakePkt))
	}

	// Verify IPv4 ID increment
	newID := binary.BigEndian.Uint16(fakePkt[4:6])
	if newID != 0x1235 {
		t.Fatalf("expected incremented ID 0x1235, got 0x%04X", newID)
	}

	// Verify PSH flag
	if (fakePkt[33] & tcpFlagPSH) == 0 {
		t.Fatalf("expected PSH flag set")
	}

	// Verify Sequence number
	seq := binary.BigEndian.Uint32(fakePkt[24:28])
	if seq != fakeSeq {
		t.Fatalf("expected seq %d, got %d", fakeSeq, seq)
	}
}

func TestIsCloudflareIP(t *testing.T) {
	// Cloudflare ranges
	if !IsCloudflareIP(net.ParseIP("104.16.1.1")) {
		t.Errorf("expected 104.16.1.1 to be Cloudflare")
	}
	if !IsCloudflareIP(net.ParseIP("172.64.0.50")) {
		t.Errorf("expected 172.64.0.50 to be Cloudflare")
	}
	if !IsCloudflareIP(net.ParseIP("188.114.96.10")) {
		t.Errorf("expected 188.114.96.10 to be Cloudflare")
	}

	// Non-Cloudflare
	if IsCloudflareIP(net.ParseIP("8.8.8.8")) {
		t.Errorf("expected 8.8.8.8 to not be Cloudflare")
	}
	if IsCloudflareIP(net.ParseIP("142.250.190.46")) {
		t.Errorf("expected Google IP to not be Cloudflare")
	}
	if IsCloudflareIP(net.ParseIP("192.168.1.1")) {
		t.Errorf("expected private IP to not be Cloudflare")
	}
}

func TestCleanDomain(t *testing.T) {
	if CleanDomain("https://cloudflare.com/path?query=1") != "cloudflare.com" {
		t.Errorf("unexpected cleaned domain for URL")
	}
	if CleanDomain("  HTTP://SpeedTest.net:443/ ") != "speedtest.net" {
		t.Errorf("unexpected cleaned domain for HTTP with port")
	}
	if CleanDomain("# this is a comment") != "" {
		t.Errorf("comment line should return empty string")
	}
	if CleanDomain("invalidhost") != "" {
		t.Errorf("domain without dot should be rejected")
	}
}

func TestKillSwitchCoordinator(t *testing.T) {
	coord := NewKillSwitchCoordinator(true)
	coord.ArmedSNI = true
	coord.ArmedCore = true

	// Heartbeat while both running -> no block
	block, unblock := coord.OnHeartbeat(true, true)
	if block || unblock || coord.IsBlocked {
		t.Fatalf("heartbeat should not trigger block when both running")
	}

	// Core crashes -> trigger block
	block, _ = coord.OnHeartbeat(true, false)
	if !block || !coord.IsBlocked {
		t.Fatalf("heartbeat should trigger block when core dropped")
	}

	// Subsequent heartbeat while down -> no double trigger
	block, _ = coord.OnHeartbeat(true, false)
	if block {
		t.Fatalf("should not double trigger block")
	}

	// Disarm both -> restore unblock
	coord.Disarm(true, false)
	unblock = coord.Disarm(false, true)
	if !unblock || coord.IsBlocked {
		t.Fatalf("disarming all should restore unblock")
	}

	// Verify Netsh args
	blockArgs := NetshBlockRuleArgs("TestRule")
	if len(blockArgs) < 5 || blockArgs[4] != "name=TestRule" {
		t.Fatalf("unexpected block args: %v", blockArgs)
	}

	unblockArgs := NetshUnblockRuleArgs("TestRule")
	if len(unblockArgs) < 5 || unblockArgs[4] != "name=TestRule" {
		t.Fatalf("unexpected unblock args: %v", unblockArgs)
	}
}

File: src/packages/lumicore/tests/sni_spoof_engine_test.rs
//! Unit tests for SNI spoof engine and DPI desync coordinator.

use lumicore::evasion::sni_spoof_engine::*;
use std::net::Ipv4Addr;

#[test]
fn test_4tuple_key_reversal() {
    let src = Ipv4Addr::new(192, 168, 1, 100);
    let dst = Ipv4Addr::new(104, 21, 5, 5);
    let key = Conn4TupleKey::new(src, 54321, dst, 443);

    assert_eq!(key.src_port, 54321);
    assert_eq!(key.dst_port, 443);

    let rev = key.reverse();
    assert_eq!(rev.src_ip, key.dst_ip);
    assert_eq!(rev.dst_ip, key.src_ip);
    assert_eq!(rev.src_port, 443);
    assert_eq!(rev.dst_port, 54321);
}

#[test]
fn test_tls_clienthello_template_synthesis() {
    let fake_sni = "speedtest.net";
    let random = [0x11u8; 32];
    let session_id = [0x22u8; 32];
    let key_share = [0x33u8; 32];

    let packet = TlsClientHelloTemplate::build(fake_sni, random, session_id, key_share);
    // Canonical length must strictly equal 517 bytes
    assert_eq!(packet.len(), 517);

    // Verify TLS record header
    assert_eq!(packet[0], 0x16); // Handshake
    assert_eq!(packet[1], 0x03); // TLS 1.0 record version
    assert_eq!(packet[2], 0x01);

    // Verify embedded random entropy
    assert_eq!(&packet[11..43], &random);
    // Verify embedded session_id
    assert_eq!(&packet[44..76], &session_id);

    // Test with another SNI of different length (e.g. 6 chars "mci.ir")
    let packet2 = TlsClientHelloTemplate::build("mci.ir", random, session_id, key_share);
    assert_eq!(packet2.len(), 517);
}

#[test]
fn test_iptcp_packet_helpers_and_fake_payload() {
    // Construct a synthetic 40-byte IPv4 + TCP ACK packet
    let mut raw_packet = vec![0u8; 40];
    // IPv4 Header (20 bytes)
    raw_packet[0] = 0x45; // Version 4, IHL 5 (20 bytes)
    raw_packet[1] = 0x00; // DSCP/ECN
    raw_packet[2] = 0x00; // Total Length (40 bytes)
    raw_packet[3] = 0x28;
    raw_packet[4] = 0x12; // Identification = 0x1234
    raw_packet[5] = 0x34;
    raw_packet[6] = 0x40; // Flags (Don't Fragment)
    raw_packet[7] = 0x00;
    raw_packet[8] = 64;   // TTL
    raw_packet[9] = 6;    // Protocol: TCP
    raw_packet[12..16].copy_from_slice(&[192, 168, 1, 10]); // Src IP
    raw_packet[16..20].copy_from_slice(&[104, 21, 5, 5]);   // Dst IP
    IpTcpPacket::compute_ip_checksum(&mut raw_packet[..20]);

    // TCP Header (20 bytes)
    raw_packet[20] = 0xD4; // Src Port: 54321 (0xD431)
    raw_packet[21] = 0x31;
    raw_packet[22] = 0x01; // Dst Port: 443 (0x01BB)
    raw_packet[23] = 0xBB;
    raw_packet[24..28].copy_from_slice(&100_000u32.to_be_bytes()); // Seq = 100,000
    raw_packet[28..32].copy_from_slice(&200_000u32.to_be_bytes()); // Ack = 200,000
    raw_packet[32] = 0x50; // Data Offset: 5 (20 bytes)
    raw_packet[33] = TCP_FLAG_ACK; // Flags: ACK
    raw_packet[34] = 0x01; // Window size: 256
    raw_packet[35] = 0x00;
    IpTcpPacket::compute_tcp_checksum(&mut raw_packet, 20, 20);

    assert!(IpTcpPacket::is_ipv4_tcp(&raw_packet));
    assert_eq!(IpTcpPacket::ip_header_len(&raw_packet), 20);
    assert_eq!(IpTcpPacket::tcp_header_len(&raw_packet, 20), 20);
    assert_eq!(IpTcpPacket::src_port(&raw_packet, 20), 54321);
    assert_eq!(IpTcpPacket::dst_port(&raw_packet, 20), 443);
    assert_eq!(IpTcpPacket::seq_num(&raw_packet, 20), 100_000);
    assert_eq!(IpTcpPacket::flags(&raw_packet, 20), TCP_FLAG_ACK);

    // Build fake payload packet
    let fake_payload = b"FAKE_TLS_CLIENT_HELLO_PAYLOAD";
    let fake_seq = 99_000u32; // (syn_seq + 1 - fake_len)
    let fake_pkt = IpTcpPacket::build_fake_payload_packet(&raw_packet, fake_payload, fake_seq)
        .expect("Build fake packet should succeed");

    assert_eq!(fake_pkt.len(), 40 + fake_payload.len());
    // Verify total length field in IPv4 header
    assert_eq!(IpTcpPacket::total_len(&fake_pkt), (40 + fake_payload.len()) as u16);
    // Verify IPv4 ID incremented from 0x1234 to 0x1235
    let new_id = u16::from_be_bytes([fake_pkt[4], fake_pkt[5]]);
    assert_eq!(new_id, 0x1235);
    // Verify TCP flags now include PSH
    assert_eq!(IpTcpPacket::flags(&fake_pkt, 20), TCP_FLAG_ACK | TCP_FLAG_PSH);
    // Verify TCP sequence overwritten
    assert_eq!(IpTcpPacket::seq_num(&fake_pkt, 20), fake_seq);
}

#[test]
fn test_cloudflare_cidr_matcher() {
    // 104.16.0.0/13
    assert!(CloudflareCidrMatcher::is_cloudflare_ip(Ipv4Addr::new(104, 16, 12, 34)));
    assert!(CloudflareCidrMatcher::is_cloudflare_ip(Ipv4Addr::new(104, 21, 88, 99)));
    // 172.64.0.0/13
    assert!(CloudflareCidrMatcher::is_cloudflare_ip(Ipv4Addr::new(172, 64, 1, 1)));
    // 162.158.0.0/15
    assert!(CloudflareCidrMatcher::is_cloudflare_ip(Ipv4Addr::new(162, 159, 10, 20)));
    // 188.114.96.0/20
    assert!(CloudflareCidrMatcher::is_cloudflare_ip(Ipv4Addr::new(188, 114, 97, 1)));

    // Non-Cloudflare public IPs
    assert!(!CloudflareCidrMatcher::is_cloudflare_ip(Ipv4Addr::new(8, 8, 8, 8)));
    assert!(!CloudflareCidrMatcher::is_cloudflare_ip(Ipv4Addr::new(142, 250, 190, 46))); // Google
    assert!(!CloudflareCidrMatcher::is_cloudflare_ip(Ipv4Addr::new(10, 0, 0, 1)));       // Private
}

#[test]
fn test_sni_domain_cleaner() {
    assert_eq!(
        SniDomainCleaner::clean_domain("https://cloudflare.com/path?q=1"),
        Some("cloudflare.com".to_string())
    );
    assert_eq!(
        SniDomainCleaner::clean_domain("   HTTP://SpeedTest.NET:443/  "),
        Some("speedtest.net".to_string())
    );
    assert_eq!(
        SniDomainCleaner::clean_domain("sub.domain.example.com:8443"),
        Some("sub.domain.example.com".to_string())
    );

    // Comment and invalid lines
    assert_eq!(SniDomainCleaner::clean_domain("# comment line"), None);
    assert_eq!(SniDomainCleaner::clean_domain("   "), None);
    assert_eq!(SniDomainCleaner::clean_domain("singleword"), None);
}

#[test]
fn test_killswitch_coordinator() {
    let mut coord = KillSwitchCoordinator::new(true);

    // Arming SNI and V2Ray
    coord.arm_sni();
    coord.arm_v2ray();

    // While both running, tick produces NoChange
    assert_eq!(coord.on_tick(true, true), KillSwitchAction::NoChange);
    assert!(!coord.is_blocked);

    // If SNI engine crashes (sni_running=false), tick produces BlockOutbound
    assert_eq!(coord.on_tick(false, true), KillSwitchAction::BlockOutbound);
    assert!(coord.is_blocked);

    // Subsequent tick while still down produces NoChange (already blocked)
    assert_eq!(coord.on_tick(false, true), KillSwitchAction::NoChange);

    // Disarm SNI and V2Ray restores unblock
    coord.disarm_sni();
    let action = coord.disarm_v2ray();
    assert_eq!(action, KillSwitchAction::UnblockOutbound);
    assert!(!coord.is_blocked);

    // Netsh rule syntax validation
    let block_cmd = KillSwitchCoordinator::netsh_block_rule("LumiNet KillSwitch");
    assert!(block_cmd.contains(&"action=block".to_string()));
    assert!(block_cmd.contains(&"name=LumiNet KillSwitch".to_string()));

    let unblock_cmd = KillSwitchCoordinator::netsh_unblock_rule("LumiNet KillSwitch");
    assert!(unblock_cmd.contains(&"delete".to_string()));
}
